### ISSUE TITLE
[FR][jit] torch.jit.script as a class decorator

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#25066]
If my module is supposed to be scripted, then it is natural to have the script wrapper close to the class definition, similar to scripting a function. Having to call `torch.jit.script` after instance creation is unintuitive, and often breaks abstraction.

So, I think `torch.jit.script` should also work as a class decorator, returning a wrapped module class that always gives the scripted module.

Some trickiness include:
+ need to correct set `__module__`, `__name__` etc. similar to `functools.wraps`.
+ probably want to have `instance` check work.

**Comment by driazati** [comment]
There's a PR for this in #22328, but still needed some more discussion before merging

**Comment by narfanar** [comment]
The current documentation makes it seem like you can just say `jit.export` instead of the old `jit.ScriptModule` & `jit.script_method` duo, but it's not clear whether that's _supposed_ to work. It _isn't_.

**Comment by driazati** [comment]
Hey @narfanar, it should be working fine, can you open a new issue with a repro for your specific issue? 

Also which part of the docs are you referring to? We’d like to make sure it’s as clear as possible

**Comment by Evpok** [comment]
@driazati They are probably referring to the fact that just adding `@jit.export` to a method won't compile it if the module is not passed to `jit.script`. `ScriptModule` had the advantage that all class instances where automatically scripted without having to call `jit.script` on them and it isn't clear how to do that with the new interface.

As for why that is useful: it allows to easily script sub-modules without scripting their parents

```python
class SubModule(jit.ScriptModule):
    […]

class MainModule(nn.Module):
    spam: Submodule

m = MainModule()
```

then `m.spam` is scripted even though `m` is not. With the new interface I have to `m.spam = jit.script(m.spam)` which is inconvenient when you have complex architectures and breaks encapsulation (not sure if this is the right word, but monkey patching class attributes this way looks fishy)

**Comment by reuben** [comment]
> With the new interface I have to `m.spam = jit.script(m.spam)` which is inconvenient when you have complex architectures and breaks encapsulation (not sure if this is the right word, but monkey patching class attributes this way looks fishy)

Maybe the better term is it breaks separation of concerns. If `SubModule` is always meant to be scripted but is used by 20 different classes, with the new interface you have to make 20 different changes outside of the definition of `SubModule` itself. If you forget one of those, you'll only notice when something goes wrong.

Maybe an alternative is to script everything *inside* `SubModule` and just leave it as an `nn.Module`? Don't know/remember if that breaks any use cases.

**Comment by narfanar** [comment]
> separation of concerns [...] always meant to be scripted

Exactly :'D

> Maybe an alternative is to script everything inside SubModule and just leave it as an nn.Module? Don't know/remember if that breaks any use cases.

The problem here: you actually can't script anything in a module, not individually.

1. `jit.script` expects either a class (not inheriting from `nn.Module`) or a plain function (not a method or a class method, though I _think_ static methods work).
    1. Using `script` on a method requires annotating the `self` parameter, which is impossible. And doesn't make sense theoretically (a TorchScript parameter must be of a TorchScript type, which means `self` must be a `ScriptModule`, so...).
2. Refactoring a module into a `ModuleDict` plus a bunch of free standing scripted functions doesn't usually work.
    1. I think `ModuleDict` is broken in scripting? I can't subscript it.
3. To work around (1) and (2), you can apply `jit.script` to a class... Except, it can't be a `nn.Module`. It has to be a plain class, because `script` compiles it to a `ScriptModule`. What does this mean? A lot of restrictions on how the class overall is designed, but they're not very clear.
    1. The most clear part for me is: `export f`: you have to obey everything within f; `unused f`: you can write f as you like.
    2. But also, a bit importantly for me, you lose tool help. No auto complete or whatever. And classes scripted that way are not recognized by `Jedi` to be (eventually?) subclasses of `ScriptModule`.

**Comment by suo** [comment]
Seems clear from this discussion that there's a gap in the API here. Let me attempt to summarize what we've so far:
- The recursive-style scripting API does not allow for modules that are scripted-on-construction.
- One obvious extension would be to allow `@torch.jit.script` to decorate an `nn.Module` class definition, just like with functions (and script classes).
- The problem with that approach is it violates the assumption that everything covered by `@torch.jit.script` is 100% scripted.

One idea is to create a new class decorator (like `@torch.jit.compile` as @SsnL has [proposed](https://github.com/pytorch/pytorch/pull/22328#issuecomment-529210455)). I have a few reservations about this:
- A new API will be potentially confusing (how is it different from `@script`?) and non-discoverable
- We will run into the same problems customizing method compilation that `@export`, `@unused`, and `@script_method` are all supposed to solve.

My proposal: lets just stop trying to kill "inheriting from `ScriptModule`" as an API and bless it for this case. Now that the inheritance-style pathway is the implementation as the recursive-style pathway it's won't actually be that hard to maintain. Afaik, there are only two behavioral differences that we need to smooth over:
- `forward()` is not automatically scripted with the inheritance-style API. Adding this will be a breaking change but a pretty "small" one, since people are probably already scripting their `forward()`.
- `@torch.jit.script_method` should be deprecated or at least made an alias for `@torch.jit.export`.

Thoughts? @driazati, @SsnL, @zdevito

**Comment by SsnL** [comment]
@suo Thanks for the detailed explanation!

Keeping `ScriptModule` around would be an okay workaround. However, it would be nice to have the consistency of always using decorators for JIT. 

My suggestion: since `ScriptModule` seems to imply the semantics of "not scripting `__init__`, only scripting `forward`", maybe we can add `torch.jit.script_module` as the decorator, which under the hood simply calls `script` on the created module object, and obeys the rules specified by `export ` and `unused`? :) 

P.S. Another major advantage of the inheritance style is that it allows `isinstance` relations (I suppose...). But maybe regular script modules should also be subclasses of the original class (https://github.com/pytorch/pytorch/issues/29092).

**Comment by Evpok** [comment]
Also in 1.3, `ScriptModule` can have an unscripted `forward` but scripted methods, would there be a way to keep that with this new API?

**Comment by zdevito** [comment]
I don't have a very strong opinion here, but my weaker opinion is that adding a way to make a script function pre-compiled will lead people to bad structure, where people will struggle to debug script code. We almost always want to encourage users to write code in such a way that both the scripted and unscripted versions are easily accessible, and having script work on instances of callable things encourages that. A decorator for class instances makes it hard to access the non-script versions. I don't find the function/module distinction of torch.jit.script to be a discrepancy because both functions and module _instances_ are the callable thing actually being used. Module _classes_ are constructors for the callable, so it is not a given then should also work with script. Furthermore, if you really want to force people to use the jit version in multiple locations, which is currently the rare case, it is still possible in a few lines:

```
class FooImpl:
   ....
class Foo:
  def __init__(*args, **kwargs): 
     self.impl = torch.jit.script(FooImpl(*args, **kwargs))
  def forward(x):
    return self.impl(x)
```
My concern would be that trying to solve isinstance trickiness (e.g. hijacking isinstance in our implementation), will lead to less understandable behavior in corner cases, so it would be better for someone to op into wrapping constructors explicitly.

**Comment by mohammedayub44** [comment]
@suo  Has this been changed already in PyTorch 1.5.0 ? 
I was using this decorator (`@torch.jit.script`) on top of a `nn.Module` Class. It did not give me error. Or do you say its better to use `@torch.jit.export` for Class of type `nn.Module`

Thanks !

**Comment by randolf-scholz** [comment]
@mohammedayub44 I found the following solution by writing a simple class wrapper


```python
def autojit(class_):

    @functools.wraps(class_, updated=())
    class WrappedClass(class_):
        
        def __new__(cls, *args, **kwargs):
            instance = class_(*args, **kwargs)
            return jit.script(instance)

    return WrappedClass
```

@SsnL  I believe the functionality proposed here could be added in a very simple manner to `jit.script` by simply checking if the input is a `type`:

```python
def patched_jit(obj):
    
    if isinstance(obj, type):
        
        @functools.wraps(obj, updated=())
        class WrappedClass(obj):
            def __new__(cls, *args, **kwargs):
                instance = obj(*args, **kwargs)
                return jit.script(instance)
        return WrappedClass
    
    return jit.script(obj)
```

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
